### DISCUSSION TITLE
Split an edge into several segments for live traffic?

### DISCUSSION THREAD

**Discussion by mogita** [valhalla/valhalla#3945] [Q&A] [answered]
I've been tweaking the code from this [live traffic POC](https://github.com/alinmindroc/valhalla_traffic_poc) to make it towards a production approach. One question I can't figure out alone is: since [TrafficSpeed](https://github.com/valhalla/valhalla/blob/master/valhalla/baldr/traffictile.h#L53) contains 2 breakpoints fields to specify the portion that the speed should apply to on an edge, how can I split an edge into several segments and apply different traffic speed within `traffic.tar`? The data modeling seems obscure to me.

Let me elaborate on the scenario with an example. Supposing I have a good data source telling me that segment 1 has an accident causing the traffic flow into a stasis, while keeping driving into segment 2 you'll have a normal speed. This could happen in real-time traffic use cases when we try to reflect the traffic accurately in the routing results. Here's a graph illustrating this:

```
|------------------ edge A -------------------|
|   segment 1     |            segment 2      |
|   speed: 12     |            speed: 60      |
```

While examining the functions [build_live_traffic_data](https://github.com/alinmindroc/valhalla_traffic_poc/blob/main/valhalla_code_overwrites/src/mjolnir/valhalla_traffic_demo_utils.cc#L128) and [customize_live_traffic_data](https://github.com/alinmindroc/valhalla_traffic_poc/blob/main/valhalla_code_overwrites/src/mjolnir/valhalla_traffic_demo_utils.cc#L188) (which originally came from [`test/test.cc`](https://github.com/valhalla/valhalla/blob/master/test/test.cc)), both wrote as many edges as `tile.header->directed_edge_count` to the archive, but without specifying an edge ID. I'm aware that the breakpoints might be used here, but still confused about how to model the data to present multiple segments.

Based on the scenario above, is it possible to take various live traffic segments into effect on a same edge?

**Comment by kevinkreiser** [comment, marked as answer]
yes. each edge can have up to 3 segments depending on how the break points are assigned. in your case you only need to use one breakpoint in edge A. the breakpoint tells where along the edge the two segments are split. read over the traffic tile structure in the source to get a better understanding of how the traffic on an edge is modeled. in our testing you rarely needed more than 3 segments so we made this the max per edge.

the reason there are no edge ids is because each one is a fixed size structure and so you can think of it as a parallel array to the array of regular edge. so if you have edge 10 to get its traffic you just grab the 10th item in its traffic array. each item represents traffic on a single edge. and each edge can have max 3 segments defined by the 2 breakpoints in the traffic record structure.

when im back to my computer I can give you an example of how to model the scenario you mentioned

**Reply by mogita** [reply]
Greatly appreciated! This has corrected my mind sharply on the meaning of the breakpoints. I thought they define 2 points of an arbitrary extent on an edge.

So based on my new understanding, do the three `encoded_speed` and `congestion` fields reflect the max 3 segments model? And `encoded_speed2` or `encoded_speed3` would be ignored when breakpoints are both set to `255`? Am I taking it correct?

**Reply by kevinkreiser** [reply]
yep that's right

**Reply by mogita** [reply]
Great, thanks for the confirmation. Then I think we need to pull an average speed value on an edge without using the breakpoints. Becuase max at 3 segments might be too few. We have the traffic data source at a per OSM node pair level, so an edge can sometimes contain up to dozens of node pairs with different speeds, making it a bit complicated to design a method to properly render them into at most 3 segments.

May I know if in your opinion, in the case of Valhalla, it is necessary to make an edge segmentation result as accurate as possible to reflect so many node pairs? Does it help with a higher ETA accuracy in general?
